### ISSUE TITLE
Unexpected results related to creation/initialization of layers and optimizer (when step function does nothing)

### ISSUE THREAD

**Issue by Kwonsoo** [LaurentMazare/ocaml-torch#36]
Hi,

I borrowed the example code in ocaml-torch/src/tests/optimization_tests.ml, and changed a little bit and ran it to see if optimization works. The code is below:
```
module T = Torch
let main () =
  Torch_core.Wrapper.manual_seed 42;
 
  let batch_size = 1 in
  let learning_rate = 1e-3 in
  let xs = T.Scalar.float 1.0 |> T.Tensor.mul1 (T.Tensor.ones [batch_size; 2]) in
  let ys = T.Tensor.f 10.0 in
  let device = T.Device.cuda_if_available () in
  let vs = T.Var_store.create ~name:"neural-analysis" ~device () in
    
  (* TWO LINES BELOW *)
  let optimizer = T.Optimizer.sgd vs ~learning_rate in  (* (1) *)
  let linear = T.Layer.linear vs ~input_dim:2 1 in  (* (2) *)
  
  for _ = 1 to 100 do
    T.Optimizer.zero_grad optimizer;
    let ys_ = T.Layer.forward linear xs in
    let loss = T.Tensor.((ys_ - ys) * (ys_ - ys)) in
    Printf.eprintf "--> %f\n" (loss |> T.Tensor.float_value);
    T.Tensor.backward loss;
    T.Optimizer.step optimizer
  done
```
When I ran this code, it kept printing the same, initial loss value.

Actually the code I am working on is more complex than this example, and it took several days to figure out what's happening, since I thought it was because I did something wrong with my domain-specific part of the code.

The problem was that I created `optimizer` BEFORE creating `linear` in the example code above. When creating an optimizer, the procedure registers trainable parameters stored in variable store `vs`, but the trainable parameters are registered at `vs` only when we create and initialize layers. So, what I should have done is create layers first (so that their trainable parameters are registered at `vs` first) and create an optimizer later to properly grab the parameters (corresponding to layers) already listed in `vs`.

The two lines ((1) and (2) in the example code) are seemingly (and syntactically) not dependent, but there is an important dependency between the two from the perspective of the whole optimization process inside the implementation. This is not an issue in Pytorch, since we must plug in parameters explicitly (e.g., `torch.optim.Adam(model.parameters(), lr=learning_rate)`) when we create an optimizer in Pytorch, thus no hidden dependency. I see that the current ocaml-torch document says nothing about this hidden dependency, and this may bother other users in the future too.

Yes, the easiest fix is to change the order of (1) and (2), and that is what I'm doing basically.
But it would help users if the document explicitly mentions this point.

Best,
Gwonsoo

**Comment by LaurentMazare** [comment]
Thanks for reporting this. This is indeed a 'well known' issue with this api, it also used to be the case with the rust api (tch-rs) until this [commit](https://github.com/LaurentMazare/tch-rs/commit/1e7f2c524a6d52425bb62c51600391f12d58545c#diff-6cb81cedee57e07fc71a41110869b55f). I will have a go at porting this fix to ocaml-torch so that this is not an issue anymore.

**Comment by Kwonsoo** [comment]
Thank you very much, Laurent.

Gwonsoo

**Comment by LaurentMazare** [comment]
I just pushed some changes so that variables can be added to the var store after creating the optimizer. Let me know if you find any issues and thanks again for raising this.

**Comment by Kwonsoo** [comment]
Thank you, Laurent.
Probably I'm not going to try the updated version soon, but I will let you know later if there are issues related to this.

Gwonsoo
